homeassistant/components/mqtt/__init__.py

```diff
@@ -1000,9 +1000,37 @@ def discovery_callback(payload):
 class MqttEntityDeviceInfo(Entity):
     """Mixin used for mqtt platforms that support the device registry."""
 
-    def __init__(self, device_config: Optional[ConfigType]) -> None:
+    def __init__(self, device_config: Optional[ConfigType],
+                 config_entry=None) -> None:
         """Initialize the device mixin."""
         self._device_config = device_config
+        self._config_entry = config_entry
+
+    async def device_info_discovery_update(self, config: dict):
+        """Handle updated discovery message."""
+        self._device_config = config.get(CONF_DEVICE)
+        device_registry = await \
+            self.hass.helpers.device_registry.async_get_registry()
+        config_entry_id = self._config_entry.entry_id
+        device_info = self.device_info
+
+        if config_entry_id is not None and device_info is not None:
+            processed_dev_info = {
+                'config_entry_id': config_entry_id
+            }
+            for key in (
+                    'connections',
+                    'identifiers',
+                    'manufacturer',
+                    'model',
+                    'name',
+                    'sw_version',
+                    'via_hub',
+            ):
+                if key in device_info:
+                    processed_dev_info[key] = device_info[key]
+
+            device_registry.async_get_or_create(**processed_dev_info)
 
     @property
     def device_info(self):
```

Comment:
This line can be removed after this

Reply:
Why is it ok to remove the device_info function?

Reply:
Not all of it. But everything except for identifiers/connections can be removed as now that's handled above.

Reply:
OK, I think it's fixed now.

Reply:
The lines are still there 🌚 
https://github.com/home-assistant/home-assistant/pull/19540/files#diff-745593be832eda57f1414e7948e5c158R1037

Reply:
Yes, but the code is now no longer duplicated in `MqttEntityDeviceInfo.device_info_discovery_update()`, so shouldn't be removed from `MqttEntityDeviceInfo.device_info()`.